### PR TITLE
feat(symmetric): support SPD GPU

### PR DESCRIPTION
**Pull request by JieRen98**
This PR
- [x] Symmetric struct
- [x] Symmetric GPU solver
- [ ] Reformat
- [ ] Symmetric matching support
- [ ] Symmetric equilibration support

Future
- [ ] Symmetric MAGMA support
- [ ] Symmetric distributed support
- [ ] Symmetric Low Rank support

### CODE REVIEW

File: src/sparse/fronts/CMakeLists.txt [hunk 1]
```diff
@@ -36,7 +36,12 @@ if(STRUMPACK_USE_CUDA OR STRUMPACK_USE_HIP OR STRUMPACK_USE_SYCL)
     PRIVATE
     ${CMAKE_CURRENT_LIST_DIR}/FrontalMatrixGPU.cpp
     ${CMAKE_CURRENT_LIST_DIR}/FrontalMatrixGPU.hpp
+    ${CMAKE_CURRENT_LIST_DIR}/FrontalMatrixGPUSymmetric.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/FrontalMatrixGPUSymmetric.hpp
     ${CMAKE_CURRENT_LIST_DIR}/FrontalMatrixGPUKernels.hpp)
+
+  find_package(Eigen3 REQUIRED)
```
Comment (pghysels) [inline src/sparse/fronts/CMakeLists.txt:43]: Is this necessary here, or only for the example?

File: src/sparse/fronts/FrontFactory.cpp [hunk 2]
```diff
@@ -142,9 +143,14 @@ namespace strumpack {
       front.reset
         (new FrontalMatrixMAGMA<scalar_t,integer_t>(s, sbegin, send, upd));
 #else
-#if defined(STRUMPACK_USE_GPU)
-      front.reset
-        (new FrontalMatrixGPU<scalar_t,integer_t>(s, sbegin, send, upd));
+#if defined(STRUMPACK_USE_CUDA) || defined(STRUMPACK_USE_HIP)
```
Comment (pghysels) [inline src/sparse/fronts/FrontFactory.cpp:146]: now you can use `STRUMPACK_USE_GPU` which also includes `STRUMPACK_USE_SYCL`

File: src/StrumpackOptions.hpp [hunk 3]
```diff
@@ -749,7 +749,17 @@ namespace strumpack {
...
 
-    /**
+//      /**
+//     * Disable symmetric solver.
+//     */
+//      void disable_symmetric() { use_symmetric_ = false; }
+      /**
+      * Disable symmetric solver.
+      */
+      void enable_symmetric() { use_symmetric_ = true; }
```
Comment (pghysels) [inline src/StrumpackOptions.hpp:759]: Should it be 'symmetric positive definite' (SPD)?

**Comment by GitHubbeer** [comment]
I observed that the matching process in STRUMPACK relies on the mc64ad algorithm (STRUMPACK ./src/sparse/MC64ad.cpp), which unfortunately does not support symmetric matrices. I found an alternative approach for Cholesky decomposition of symmetric matrices by incorporating the mc64_matching from HSL_MC64 ("matrix type indicates if the matrix is symmetric or not".  section 2.5 in https://www.hsl.rl.ac.uk/specs/hsl_mc64_C.pdf). So I think we should implement our own symmetric version or use HSL_MC64 directly. I suggest that we can make it a separate pull request.

**Comment by GitHubbeer** [comment]
@pghysels Could you please check this PR?

**Comment by pghysels** [comment]
Can you add a few tests?
You can run `testSymmetricPositiveDefinite` and `testMixedPrecisionSymmetricPositiveDefinite.cpp` in `test/CMakeLists.txt`.
You can download a small SPD matrix in `download_mtx.sh`.

Maybe you can rename files like `testSymmetricPositiveDefinite` to `testSPD`.

**Comment by pghysels** [comment]
Maybe we should add a routine like `set_matrix` that takes a symmetric matrix, i.e., only taking the lower/upper triangle.

The current `CSRMatrix` class (and `CompressedSparseMatrix`) have a `symm_sparse` argument. But this flag was only meant to denote whether the sparsity pattern is symmetric. It is not about the value symmetry or positive definiteness. So I feel like we should not confuse these things.

**Comment by JieRen98** [comment]
How did you format codes? @pghysels

**Comment by pghysels** [comment]
> How did you format codes? @pghysels

I don't use any strict formatting guidelines.
We should probably setup auto formatting using clang-format or something like that.

**Comment by JieRen98** [comment]
> > How did you format codes? @pghysels
> 
> I don't use any strict formatting guidelines. We should probably setup auto formatting using clang-format or something like that.

Sounds great! I opened a new issue (https://github.com/pghysels/STRUMPACK/issues/114) for this. I think we can do this in the next PR for clarity because I usually follow the philosophy that one PR for one feature.

**Comment by GitHubbeer** [comment]
"test_SPD_mixedPrecison" and "test_SPD_seq" in test/CMakeList and routine "set_lower_triangle_matrix" for taking a symmetric matrix are ready. Please review. I also changed symm_sparse() to is_symmetric(opts_) to check whether the matrix is symmetric.

**Comment by pghysels** [comment]
This looks good. I will merge and then clean up a bit myself.

There might be licensing issue with the newer MC64 version.

**Pull request merged**